File: JarvisClient/Models/ProjectItemKind.cs
namespace JarvisClient.Models;

[AutoClosed(true)]
public partial record ProjectItemKind
{
    partial record ProjectFile(string Name, long FileSize, DateTimeOffset CreationDate, DateTimeOffset ModificationDate);

    partial record ProjectFolder(string Name);

    partial record ProjectFileError(string Name, string ErrorMessage);

    static partial class Cons
    {
        public static ProjectItemKind NewProjectFile(FileInfo info) =>
            NewProjectFile(info.Name, info.Length, info.CreationTime, info.LastWriteTime);
    }
}

File: Client/ProjectItemKind.fs
namespace Client

open System
open System.IO
open System.Text.Json.Serialization
open FsCodec.SystemTextJson

[<JsonConverter(typeof<UnionConverter<ProjectItemKind>>)>]
type ProjectItemKind =
    | ProjectFile of name: string * fileSize: int64 * creationDate: DateTimeOffset * modificationDate: DateTimeOffset
    | ProjectFolder of name: string
    | ProjectFileError of name: string * errorMessage: string

module ProjectItemKind =
    let ofFileInfo (info: FileInfo) =
        ProjectFile(info.Name, info.Length, DateTimeOffset info.CreationTime, DateTimeOffset info.LastWriteTime)
